tools/ci_build/rewriter.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import os
+import shutil
+import onnx
+from onnx import AttributeProto as AP
+
+#pylint: disable=no-member,too-many-locals,too-many-statements
+
+def extract_ops_from_file(file_path, referred_ops):
+    '''extract ops from csv file with format:
+        op_type,domain,opset'''
+
+    if not os.path.isfile(file_path):
+        return referred_ops
+
+    with open(file_path, 'r') as csv_to_read:
+
+        for line in csv_to_read.readlines():
+            op_type, domain, raw_opset = line.strip().split(',')
+            opset = int(raw_opset)
+
+            if op_type not in referred_ops:
+                referred_ops[op_type] = {domain:[opset]}
+
+            elif domain not in referred_ops[op_type]:
+                referred_ops[op_type][domain] = [opset]
+
+            elif opset not in referred_ops[op_type][domain]:
+                referred_ops[op_type][domain].append(opset)
+
+    return referred_ops #end of extract_ops_from_file(...)
+
+
+def extract_ops_from_model(model_path, referred_ops):
+    '''extract ops from models under model_path and return a diction'''
+
+    if not os.path.isdir(model_path):
+        return referred_ops
+
+    def map_domain(domain):
+        if domain == 'ai.onnx.ml':
+            return 'kMLDomain'
+        return 'kOnnxDomain'
+
+    def extract_ops_from_graph(graph, opsets, operators):
+        '''extract ops from graph and all subgraphs'''
+
+        for operator in graph.node:
+
+            if operator.op_type not in operators:
+                operators[operator.op_type] = {}
+
+            for opset in opsets:
+                mapped_domain = map_domain(opset.domain)
+                if mapped_domain not in operators[operator.op_type]:
+                    operators[operator.op_type][mapped_domain] = []
+                if opset.version not in operators[operator.op_type][mapped_domain]:
+                    operators[operator.op_type][mapped_domain].append(opset.version)
+
```

Comment:
why do we repeat this for every node? opsets is per graph isn't it?

Reply:
Yeah but we are dealing with multiple models at one time right?

Reply:
You're doing this inside the loop of the nodes.

---
In reply to: [468920816](https://github.com/microsoft/onnxruntime/pull/4747#discussion_r468920816) [](ancestors = 468920816)

Reply:
So all I need is just deal with opsets[0] where opsets is from the model proto?

Reply:
No. You need to iterate them but it's once per model, so this can be moved up prior to the `for oeprator in graph.node:` loop. 

---
In reply to: [468924320](https://github.com/microsoft/onnxruntime/pull/4747#discussion_r468924320) [](ancestors = 468924320)

Reply:
Done.